compiler/rustc_middle/src/traits/mod.rs

```diff
@@ -962,3 +962,21 @@ pub enum MethodViolationCode {
     /// the method's receiver (`self` argument) can't be dispatched on
     UndispatchableReceiver,
 }
+
+/// These are the error cases for `codegen_fulfill_obligation`.
+#[derive(Copy, Clone, Debug, Hash, HashStable, Encodable, Decodable)]
+pub enum CodegenObligationError {
```

Comment:
I think this is the best change -- we have far too many cases where `codegen_fulfill_obligation` can fail (mostly with trivially unsatisfiable bounds) that treating the query as infallible gives us a lot of ICEs, even without the MIR inline pass.